### PR TITLE
feat: rollback workaround #13347, improve stack traces using async stack traces instead

### PR DESCRIPTION
**Pull request by ephys**
## Pull Request Checklist

- [ ] Have you added new tests to prevent regressions?
- [ ] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? 
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description Of Change

Fixes https://github.com/sequelize/sequelize/issues/14807
Closes https://github.com/sequelize/sequelize/pull/15348

This PR removes the stack trace modification introduced in #13347. Instead, I've rewritten the necessary Query implementations to use an async flow that is compatible with native async stack traces. Resulting in the clean stacktraces posted here: https://github.com/sequelize/sequelize/pull/15555#issuecomment-1383136934

I think this can be backported to v6 too, there shouldn't be any breaking change (apart from the use of private methods which may not supported in node 10)

### CODE REVIEW

File: packages/core/src/dialects/db2/query.js [hunk 1]
```diff
@@ -36,152 +36,154 @@ export class Db2Query extends AbstractQuery {
...
-        }, this.options.transaction.name);
-      } else {
-        const params = [];
-        if (parameters) {
-          _.forOwn(parameters, (value, key) => {
-            const param = this.getSQLTypeFromJsType(value, key);
-            params.push(param);
-          });
-        }
+    if (this.sql.startsWith('BEGIN TRANSACTION')) {
```
Comment (WikiRik) [inline packages/core/src/dialects/db2/query.js:39]: Not really related to this PR, but the imports of `util` and `dayjs` are unused in this file.

File: packages/core/src/dialects/ibmi/query.js [hunk 2]
```diff
@@ -15,41 +15,35 @@ export class IBMiQuery extends AbstractQuery {
...
-      this.connection.query(this.sql, parameters, (error, results) => {
+    const complete = this._logQuery(sql, debug, parameters);
 
-        if (error) {
-          const formattedError = this.formatError(error, stacktrace);
-          reject(formattedError);
-
-          return;
-        }
+    let results;
```
Comment (WikiRik) [inline packages/core/src/dialects/ibmi/query.js:22]: Not related to this PR, but the import of `_` is unused

Comment (ephys) [reply packages/core/src/dialects/ibmi/query.js:21]: Done

File: packages/core/src/dialects/db2/query.js (hunk 1 above)
Comment (ephys) [reply packages/core/src/dialects/db2/query.js:37]: Done

Comment (WikiRik) [reply packages/core/src/dialects/db2/query.js:37]: Hmm, I thought these were still pending. But apparently not

File: packages/core/src/dialects/sqlite/query.js [hunk 3]
```diff
@@ -185,87 +180,113 @@ export class SqliteQuery extends AbstractQuery {
...
+      complete();
 
-        if (this.getDatabaseMethod() === 'all') {
-          let tableNames = [];
-          if (this.options && this.options.tableNames) {
-            tableNames = this.options.tableNames;
-          } else if (/from `(.*?)`/i.test(this.sql)) {
-            tableNames.push(/from `(.*?)`/i.exec(this.sql)[1]);
-          }
+      return query._handleQueryResponse(response.statement, columnTypes, response.results);
```
Comment (WikiRik) [inline packages/core/src/dialects/sqlite/query.js:228]:
```suggestion
      return query._handleQueryResponse(response.statement, response.results);
```

File: packages/core/src/dialects/sqlite/query.js [hunk 4]
```diff
@@ -49,12 +49,7 @@ export class SqliteQuery extends AbstractQuery {
     return ret;
   }
 
-  _handleQueryResponse(metaData, columnTypes, err, results, errStack) {
-    if (err) {
-      err.sql = this.sql;
-      throw this.formatError(err, errStack);
-    }
-
+  _handleQueryResponse(metaData, columnTypes, results) {
```
Comment (WikiRik) [inline packages/core/src/dialects/sqlite/query.js:52]:
```suggestion
  _handleQueryResponse(metaData, results) {
```

**Review by WikiRik — COMMENTED** [review]
This time not as single comments 😅 

Two small comments on sqlite, other files look good. I'll look at db2 once the tests have passed

File: packages/core/src/dialects/sqlite/query.js (hunk 4 above)
Comment (ephys) [reply packages/core/src/dialects/sqlite/query.js]: Done

**Comment by ephys** [comment]
I just discovered the creation of savepoint in our db2 implementation commits the transaction and creates a new one. That's not at all what a savepoint is supposed to do, we'll need to fix that, and it's something to take into account for the default value of https://github.com/sequelize/sequelize/issues/13741

**Review by WikiRik — APPROVED** [review]

**Pull request merged**